keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1763 @@
+containers in a pod. Depending on the capabilities defined in the
+ResourceClaim by the driver, a ResourceClaim can be used exclusively
+by one pod at a time, by a certain maximum number of pods, or an
+unlimited number of pods. Kubernetes can only count, but not do more
+than that because it has no understanding of what other limitations might
+apply to specific resources. In practice, one or unlimited are
+expected to be the common cases. A fixed number is supported by the API
+because it is easy to implement and might be
+useful.
+
+Users of a ResourceClaim don't need to be Pods. This KEP specifically supports
+Pods as users and describes how kube-scheduler and kubelet will deal with Pods
+that depend on a ResourceClaim, but the API and some custom resource driver
+might also be useful for controllers to manage resources without using those
+resources for Pods.
+
+### User Stories
+
+#### Cluster add-on development
+
+As a hardware vendor, I want to make my hardware available also to applications
+that run in a container under Kubernetes. I want to make it easy for a cluster
+administrator to configure a cluster where some nodes have this hardware.
+
+I develop two components, one that runs as part of the Kubernetes control plane
+and one that runs on each node, and package those inside container images. YAML
+files describe how to deploy my software on a Kubernetes cluster that supports
+dynamic resource allocation.
+
+Documentation for administrators explains how the nodes need to be set
+up. Documentation for users explains which parameters control the behavior of
+my hardware and how to use it inside a container.
+
+#### Cluster configuration
+
+As a cluster administrator, I want to make GPUs from vendor ACME available to users
+of that cluster. I prepare the nodes and deploy the vendor's components with
+`kubectl create`.
+
+I create a ResourceClass for the hardware with parameters that only I as the
+administrator am allowed to choose, like for example running a command with
+root privileges that does some cluster-specific initialization for each allocation:
+```
+apiVersion: cdi.k8s.io/v1alpha1
+metadata:
+  name: acme-gpu
+driverName: gpu.acme.com
+parameters:
+  initCommand:
+  - /usr/local/bin/acme-gpu-init
+  - --cluster
+  - my-cluster
+```
+
+#### Partial GPU allocation
+
+As a user, I want to use a GPU as accelerator, but don't need exclusive access
+to that GPU. Running my workload with just 2Gb of memory is sufficient. This is
+supported by the ACME GPU hardware. I know that the administrator has created
+an "acme-gpu" ResourceClass.
+
+For a simple trial, I create a Pod directly where two containers share the same subset
+of the GPU:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  resourceClaims:
+  - name: "gpu" # this name gets referenced below under "claims"
+    template:
+      resourceClassName: "acme-gpu"
+      parameters:
+        memory: "2Gi"
```

Comment:
based on the unconference, and just spitballing here: perhaps something like

```suggestion
   resourceClassRef:
     name: "acme"
     parameters:
       - resourceParameterRef:
           name: acme-gpu-superfast
           kind: gpuParameters
           version: resources.xy.z/v1alpha1
```

I don't really see how this works neatly (reliably?) in either the raw object form or this form. I.e. how can we validate the parameters are valid before submitting to the cluster, or at least before attempting to start the pod? Maybe it makes more sense that the ref in pod spec is to the already-created parameter group, which inherits the resourceClass. Then the controller can pre-validate or pre-create things, and admission is similar to secret/configmap refs - easy to check that it exists and status is ok. This is more or less the pattern of crossplane, but the missing link, which this KEP addresses, is how to bind those to particular pods.

Reply:
So you are suggesting that Instead of a builtin ResourceClaim type we have vendor-defined types that have some pre-defined fields and then in-tree code just works with those fields (duck typing)?

That doesn't solve the problem of validating the template of such a custom resource claim, does it?

> since the resourceClaim is persisted after the pod is gone (I think?)

No, automatically generated ResourceClaims have the Pod as owner and get garbage-collected when the Pod gets deleted. This ownership reference gets checked to ensure that a Pod uses "its" instance of the ResourceClaim and not the one from a previous incarnation of itself or some manually created ResourceClaim.